# Evaluation Exercises

### 2. Given the following confusion matrix, evaluate (by hand) the model's performance.


|               | pred dog   | pred cat   |
|:------------  |-----------:|-----------:|
| actual dog    |         46 |         7  |
| actual cat    |         13 |         34 |

- In the context of this problem, what is a false positive?
    - False Positive: Predicted it was a dog, but it was a cat
        
        
- In the context of this problem, what is a false negative?
     - False Negative: Predicted it was a cat, but it was a dog
        
        
- How would you describe this model?
     - I would say it is a pretty good model since the True Positives and True Negatives are high meaning it has high Accuracy (80% Accurate)

### 3. You are working as a datascientist working for Codeup Cody Creator (C3 for short), a rubber-duck manufacturing plant.

Unfortunately, some of the rubber ducks that are produced will have defects. Your team has built several models that try to predict those defects, and the data from their predictions can be found here.

Use the predictions dataset and pandas to help answer the following questions:

 - An internal team wants to investigate the cause of the manufacturing defects. They tell you that they want to identify as many of the ducks that have a defect as possible. Which evaluation metric would be appropriate here? Which model would be the best fit for this use case?


      - RECALL because we want to reduce False Negatives since we don't want defects found when we predicted no defects.  Cost of False Positives is more than the cost of False Negatives.

- Recently several stories in the local news have come out highlighting customers who received a rubber duck with a defect, and portraying C3 in a bad light. The PR team has decided to launch a program that gives customers with a defective duck a vacation to Hawaii. They need you to predict which ducks will have defects, but tell you the really don't want to accidentally give out a vacation package when the duck really doesn't have a defect. Which evaluation metric would be appropriate here? Which model would be the best fit for this use case?


    - PRECISION because we want to reduce False Positives since we don't want to predict no defects when we actually had a defect.  Cost of False Negatives is more than the cost of False Positives.

In [127]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [128]:
csv = "c3.csv"
df= pd.read_csv(csv)

In [129]:
df.head()

actual     model1  model2     model3
0  No Defect  No Defect  Defect  No Defect
1  No Defect  No Defect  Defect     Defect
2  No Defect  No Defect  Defect  No Defect
3  No Defect     Defect  Defect     Defect
4  No Defect  No Defect  Defect  No Defect

In [130]:
# pd.crosstab(df.actual, df.prediction)
pd.crosstab(df.model1, df.actual)

actual     Defect  No Defect
model1                      
Defect          8          2
No Defect       8        182

In [131]:
pd.crosstab(df.model2, df.actual)

actual     Defect  No Defect
model2                      
Defect          9         81
No Defect       7        103

In [132]:
pd.crosstab(df.model3, df.actual)

actual     Defect  No Defect
model3                      
Defect         13         86
No Defect       3         98

In [133]:
# Which label appears most frequently in our dataset?
df.actual.value_counts()

No Defect    184
Defect        16
Name: actual, dtype: int64

In [134]:
# create a new column called 'baseline_prediction'
# which is the most frequent label e.g. in this case 'No Defect'

df['baseline_prediction']='Defect'
df

actual     model1     model2     model3 baseline_prediction
0    No Defect  No Defect     Defect  No Defect              Defect
1    No Defect  No Defect     Defect     Defect              Defect
2    No Defect  No Defect     Defect  No Defect              Defect
3    No Defect     Defect     Defect     Defect              Defect
4    No Defect  No Defect     Defect  No Defect              Defect
..         ...        ...        ...        ...                 ...
195  No Defect  No Defect     Defect     Defect              Defect
196     Defect     Defect  No Defect  No Defect              Defect
197  No Defect  No Defect  No Defect  No Defect              Defect
198  No Defect  No Defect     Defect     Defect              Defect
199  No Defect  No Defect  No Defect     Defect              Defect

[200 rows x 5 columns]

In [135]:
# Model and baseline accuracy:

model_accuracy = (df.actual == df.prediction).mean()
baseline_accuracy = (df.actual == df.baseline_prediction).mean()

print(f'   model accuracy: {model_accuracy:.2%}')
print(f'baseline accuracy: {baseline_accuracy:.2%}')

AttributeError: 'DataFrame' object has no attribute 'prediction'

### You are working as a data scientist for Gives You Paws ™, a subscription based service that shows you cute pictures of dogs or cats (or both for an additional fee).

At Gives You Paws, anyone can upload pictures of their cats or dogs. The photos are then put through a two step process. First an automated algorithm tags pictures as either a cat or a dog (Phase I). Next, the photos that have been initially identified are put through another round of review, possibly with some human oversight, before being presented to the users (Phase II).

Several models have already been developed with the data, and you can find their results here.

Given this dataset, use pandas to create a baseline model (i.e. a model that just predicts the most common class) and answer the following questions:

In terms of accuracy, how do the various models compare to the baseline model? Are any of the models better than the baseline?
Suppose you are working on a team that solely deals with dog pictures. Which of these models would you recomend for Phase I? For Phase II?
Suppose you are working on a team that solely deals with cat pictures. Which of these models would you recomend for Phase I? For Phase II?

In [137]:
csv = "gives_you_paws.csv"
df= pd.read_csv(csv)

In [138]:
df.head()

actual model1 model2 model3 model4
0    cat    cat    dog    cat    dog
1    dog    dog    cat    cat    dog
2    dog    cat    cat    cat    dog
3    dog    dog    dog    cat    dog
4    cat    cat    cat    dog    dog

In [139]:
# pd.crosstab(df.actual, df.prediction)
pd.crosstab(df.model1, df.actual)

actual   cat   dog
model1            
cat     1423   640
dog      323  2614

In [140]:
pd.crosstab(df.model2, df.actual)

actual   cat   dog
model2            
cat     1555  1657
dog      191  1597

In [141]:
pd.crosstab(df.model3, df.actual)

actual  cat   dog
model3           
cat     893  1599
dog     853  1655

In [142]:
pd.crosstab(df.model4, df.actual)

actual   cat   dog
model4            
cat      603   144
dog     1143  3110

In [148]:
# Which label appears most frequently in our dataset?
df.actual.value_counts()

dog    3254
cat    1746
Name: actual, dtype: int64

In [149]:
# create a new column called 'baseline_prediction'
# which is the most frequent label e.g. in this case 'No Defect'

df['baseline_prediction']='dog'
df

actual model1 model2 model3 model4 baseline_prediction
0       cat    cat    dog    cat    dog                 dog
1       dog    dog    cat    cat    dog                 dog
2       dog    cat    cat    cat    dog                 dog
3       dog    dog    dog    cat    dog                 dog
4       cat    cat    cat    dog    dog                 dog
...     ...    ...    ...    ...    ...                 ...
4995    dog    dog    dog    dog    dog                 dog
4996    dog    dog    cat    cat    dog                 dog
4997    dog    cat    cat    dog    dog                 dog
4998    cat    cat    cat    cat    dog                 dog
4999    dog    dog    dog    dog    dog                 dog

[5000 rows x 6 columns]

### 4.a. In terms of accuracy, how do the various models compare to the baseline model? Are any of the models better than the baseline?

In [150]:
# Model and baseline accuracy:

model1_accuracy = (df.actual == df.model1).mean()
model2_accuracy = (df.actual == df.model2).mean()
model3_accuracy = (df.actual == df.model3).mean()
model4_accuracy = (df.actual == df.model4).mean()

baseline_accuracy = (df.actual == df.baseline_prediction).mean()

print(f'   model 1 accuracy: {model1_accuracy:.2%}')
print(f'   model 2 accuracy: {model2_accuracy:.2%}')
print(f'   model 3 accuracy: {model3_accuracy:.2%}')
print(f'   model 4 accuracy: {model4_accuracy:.2%}')
print('')

print(f'(all dog) baseline accuracy: {baseline_accuracy:.2%}')



   model 1 accuracy: 80.74%
   model 2 accuracy: 63.04%
   model 3 accuracy: 50.96%
   model 4 accuracy: 74.26%

(all dog) baseline accuracy: 65.08%


##### Model 1 and Model 4 are better than the baseline model, where Model 1 is the best.

### 4.b. Suppose you are working on a team that solely deals with dog pictures. Which of these models would you recomend for Phase I? For Phase II?

In [160]:
# Recall - calculate on subset of data
# only look at subset of actual positives (i.e actual == dog)
subset = df[df.actual == 'dog']
subset

actual model1 model2 model3 model4 baseline_prediction
1       dog    dog    cat    cat    dog                 dog
2       dog    cat    cat    cat    dog                 dog
3       dog    dog    dog    cat    dog                 dog
5       dog    dog    dog    dog    dog                 dog
8       dog    dog    cat    dog    dog                 dog
...     ...    ...    ...    ...    ...                 ...
4993    dog    dog    cat    dog    dog                 dog
4995    dog    dog    dog    dog    dog                 dog
4996    dog    dog    cat    cat    dog                 dog
4997    dog    cat    cat    dog    dog                 dog
4999    dog    dog    dog    dog    dog                 dog

[3254 rows x 6 columns]

In [161]:
# calculate recall for model and baseline
model1_recall = (subset.actual == subset.model1).mean()
baseline_recall =  (subset.actual == subset.baseline_prediction).mean()

print(f'   model 1 recall: {model1_recall:.2%}')
print(f'baseline recall: {baseline_recall:.2%}')

   model 1 recall: 80.33%
baseline recall: 100.00%


In [162]:
# calculate recall for model and baseline
model2_recall = (subset.actual == subset.model2).mean()
baseline_recall =  (subset.actual == subset.baseline_prediction).mean()

print(f'   model 2 recall: {model2_recall:.2%}')
print(f'baseline recall: {baseline_recall:.2%}')

   model 2 recall: 49.08%
baseline recall: 100.00%


In [163]:
# calculate recall for model and baseline
model3_recall = (subset.actual == subset.model3).mean()
baseline_recall =  (subset.actual == subset.baseline_prediction).mean()

print(f'   model 3 recall: {model3_recall:.2%}')
print(f'baseline recall: {baseline_recall:.2%}')

   model 3 recall: 50.86%
baseline recall: 100.00%


In [164]:
# calculate recall for model and baseline
model4_recall = (subset.actual == subset.model4).mean()
baseline_recall =  (subset.actual == subset.baseline_prediction).mean()

print(f'   model 4 recall: {model4_recall:.2%}')
print(f'baseline recall: {baseline_recall:.2%}')


   model 4 recall: 95.57%
baseline recall: 100.00%


In [165]:
# Recall because we would rather catch any dog pictures even though it was a cat. 
# I would recommend model 4 because it has a recall value of 95.57%. 

### 4.c. Suppose you are working on a team that solely deals with cat pictures. Which of these models would you recomend for Phase I? For Phase II?

In [151]:
# Precision: # only look at the subset where we made positive prediction (i.e. prediction == 'dog')

subset1 = df[df.model1 == 'dog']
subset1

actual model1 model2 model3 model4 baseline_prediction
1       dog    dog    cat    cat    dog                 dog
3       dog    dog    dog    cat    dog                 dog
5       dog    dog    dog    dog    dog                 dog
7       cat    dog    cat    cat    dog                 dog
8       dog    dog    cat    dog    dog                 dog
...     ...    ...    ...    ...    ...                 ...
4992    dog    dog    cat    cat    dog                 dog
4993    dog    dog    cat    dog    dog                 dog
4995    dog    dog    dog    dog    dog                 dog
4996    dog    dog    cat    cat    dog                 dog
4999    dog    dog    dog    dog    dog                 dog

[2937 rows x 6 columns]

In [152]:
model1_precision = (subset1.actual == subset1.model1).mean()
print(f'model 1 precision: {model1_precision:.2%}')

model 1 precision: 89.00%


In [153]:
subset2 = df[df.model2 == 'dog']
subset2

actual model1 model2 model3 model4 baseline_prediction
0       cat    cat    dog    cat    dog                 dog
3       dog    dog    dog    cat    dog                 dog
5       dog    dog    dog    dog    dog                 dog
9       dog    cat    dog    cat    dog                 dog
10      dog    dog    dog    dog    dog                 dog
...     ...    ...    ...    ...    ...                 ...
4983    dog    cat    dog    dog    dog                 dog
4988    dog    dog    dog    dog    cat                 dog
4990    dog    dog    dog    cat    dog                 dog
4995    dog    dog    dog    dog    dog                 dog
4999    dog    dog    dog    dog    dog                 dog

[1788 rows x 6 columns]

In [154]:
model2_precision = (subset2.actual == subset2.model2).mean()
print(f'model 2 precision: {model2_precision:.2%}')

model 2 precision: 89.32%


In [155]:
subset3 = df[df.model3 == 'dog']
subset3

actual model1 model2 model3 model4 baseline_prediction
4       cat    cat    cat    dog    dog                 dog
5       dog    dog    dog    dog    dog                 dog
8       dog    dog    cat    dog    dog                 dog
10      dog    dog    dog    dog    dog                 dog
13      dog    cat    dog    dog    dog                 dog
...     ...    ...    ...    ...    ...                 ...
4993    dog    dog    cat    dog    dog                 dog
4994    cat    cat    cat    dog    dog                 dog
4995    dog    dog    dog    dog    dog                 dog
4997    dog    cat    cat    dog    dog                 dog
4999    dog    dog    dog    dog    dog                 dog

[2508 rows x 6 columns]

In [156]:
model3_precision = (subset3.actual == subset3.model3).mean()
print(f'model 3 precision: {model3_precision:.2%}')

model 3 precision: 65.99%


In [157]:
subset4 = df[df.model4 == 'dog']
subset4

actual model1 model2 model3 model4 baseline_prediction
0       cat    cat    dog    cat    dog                 dog
1       dog    dog    cat    cat    dog                 dog
2       dog    cat    cat    cat    dog                 dog
3       dog    dog    dog    cat    dog                 dog
4       cat    cat    cat    dog    dog                 dog
...     ...    ...    ...    ...    ...                 ...
4995    dog    dog    dog    dog    dog                 dog
4996    dog    dog    cat    cat    dog                 dog
4997    dog    cat    cat    dog    dog                 dog
4998    cat    cat    cat    cat    dog                 dog
4999    dog    dog    dog    dog    dog                 dog

[4253 rows x 6 columns]

In [158]:
model4_precision = (subset4.actual == subset4.model4).mean()
print(f'model 4 precision: {model4_precision:.2%}')

model 4 precision: 73.12%


In [159]:
# Precision because we would rather catch more cats even if it is a dog
# I would recommend model 2 because it is more precise with a precision of 89.32%. 

### 5. Read the documentation about each function, then apply those functions to the data from the previous problem.

- a. sklearn.metrics.accuracy_score

In [76]:
 from sklearn.metrics import accuracy_score

In [77]:
 accuracy_score(df.model1, df.actual)

0.8074

In [78]:
 accuracy_score(df.model2, df.actual)

0.6304

In [79]:
 accuracy_score(df.model3, df.actual)

0.5096

In [80]:
 accuracy_score(df.model4, df.actual)

0.7426

In [81]:
 accuracy_score(df.baseline_prediction, df.actual)

0.3492

- b. sklearn.metrics.precision_score

In [82]:
from sklearn.metrics import precision_score

In [182]:
precision_score(df.actual, df.model1, average='macro')

0.7898980051430666

In [184]:
precision_score(df.actual, df.model2, average='macro')

0.6886493880609905

In [185]:
precision_score(df.actual, df.model3, average='macro')

0.5091175333635416

In [186]:
precision_score(df.actual, df.model4, average='macro')

0.7692387230558727

- c. sklearn.metrics.recall_score

In [173]:
from sklearn.metrics import recall_score


In [178]:
recall_score(df.actual,df.model1, average='macro')

0.8091623596933477

In [179]:
recall_score(df.actual,df.model2, average='macro')

0.6906938398488845

In [180]:
recall_score(df.actual,df.model3, average='macro')

0.5100297739111823

In [181]:
recall_score(df.actual,df.model4, average='macro')

0.6505537989722403

- d. sklearn.metrics.classification_report

In [166]:
from sklearn.metrics import classification_report

In [168]:
print("Model 1")
pd.DataFrame(classification_report(df.actual,df.model1, labels = ['cat', 'dog'], output_dict=True)).T

precision    recall  f1-score    support
cat            0.689772  0.815006  0.747178  1746.0000
dog            0.890024  0.803319  0.844452  3254.0000
accuracy       0.807400  0.807400  0.807400     0.8074
macro avg      0.789898  0.809162  0.795815  5000.0000
weighted avg   0.820096  0.807400  0.810484  5000.0000

In [169]:
print("Model 2")
pd.DataFrame(classification_report(df.actual,df.model2, labels = ['cat', 'dog'], output_dict=True)).T

Model 2


precision    recall  f1-score    support
cat            0.484122  0.890607  0.627269  1746.0000
dog            0.893177  0.490781  0.633479  3254.0000
accuracy       0.630400  0.630400  0.630400     0.6304
macro avg      0.688649  0.690694  0.630374  5000.0000
weighted avg   0.750335  0.630400  0.631310  5000.0000

In [170]:
print("Model 3")
pd.DataFrame(classification_report(df.actual,df.model3, labels = ['cat', 'dog'], output_dict=True)).T

Model 3


precision    recall  f1-score    support
cat            0.358347  0.511455  0.421425  1746.0000
dog            0.659888  0.508605  0.574453  3254.0000
accuracy       0.509600  0.509600  0.509600     0.5096
macro avg      0.509118  0.510030  0.497939  5000.0000
weighted avg   0.554590  0.509600  0.521016  5000.0000

In [171]:
print("Model 4")
pd.DataFrame(classification_report(df.actual,df.model4, labels = ['cat', 'dog'], output_dict=True)).T

Model 4


precision    recall  f1-score    support
cat            0.807229  0.345361  0.483755  1746.0000
dog            0.731249  0.955747  0.828560  3254.0000
accuracy       0.742600  0.742600  0.742600     0.7426
macro avg      0.769239  0.650554  0.656157  5000.0000
weighted avg   0.757781  0.742600  0.708154  5000.0000

- Cat team phase 1 is model 2 with recall 0.89
- Cat team phase 2 is model 4 with precision 0.81